### PR TITLE
Dev pa user

### PR DESCRIPTION
**Pull request by racine-p-a**
Seeding database with a first admin user (data extracted from `.env` file).

I was supposed to work on `Projects` but this needed authentication... So I am working back to user registration and authentication hence the seeding. You can already expect more seeding into the database.

Rebasing was awfully wrong and buggy so I merged instead.

### CODE REVIEW

**Review by Pipazoul — APPROVED** [review]

**Comment by lutangar** [comment]
@racine-p-a I think I know what went wrong... once a past branch has been merge we usually delete it and create new feature branch from dev, see :
https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow#:~:text=The%20Feature%20Branch%20Workflow%20assumes,work%20on%20a%20new%20feature.

File: apps/backend/src/user/user.service.ts [hunk 1]
```diff
@@ -15,7 +17,7 @@ export class UserService {
    * TODO : passwords are stored clearly in database -> hash them !
    */
   async createUser(userDto: UserDto): Promise<number> {
-    const user = new UserEntity(userDto.username, userDto.password, userDto.role, userDto.active);
+    const user = new UserEntity(userDto.username, hashString(userDto.password), userDto.role, userDto.active);
```
Comment (lutangar) [inline apps/backend/src/user/user.service.ts]: Beware, here you're not using your `hashString` function but the one imported from `'nx/src/lock-file/utils/hashing'` and it's probably not what you want !

File: apps/backend/src/user/user.service.ts [hunk 2]
```diff
@@ -39,7 +41,7 @@ export class UserService {
     const id = user.userId;
     return this.userRepository.findOne({ id }).then((dataUser) => {
       dataUser.username = user.username;
-      dataUser.password = user.password;
+      dataUser.password = hashString(user.password);
```
Comment (lutangar) [inline apps/backend/src/user/user.service.ts]: same here

File: apps/backend/src/user/user.service.ts (hunk 1 above)
Comment (lutangar) [inline apps/backend/src/user/user.service.ts]: `UserEntity` class could use some kind of `fromDto` method instead.

File: apps/backend/src/user/user.service.ts [hunk 3]
```diff
@@ -66,4 +68,26 @@ export class UserService {
...
     });
   }
+
+  async createUsersOnStartUp(
+    userAdmin: Pick<UserDto, 'username' | 'password'>,
+    userDummyEditor: Pick<UserDto, 'username' | 'password'>
+  ) {
+    const admin = new UserEntity(userAdmin.username, hashString(userAdmin.password), Roles.ADMIN, true);
+    if (await this.isUsernameAlreadyInDatabase(admin.username)) {
+      console.log('Admin already in database.');
```
Comment (lutangar) [inline apps/backend/src/user/user.service.ts]: You should us the Logger object instead and set the proper log level.

File: apps/backend/src/user/user.service.ts [hunk 4]
```diff
@@ -66,4 +68,26 @@ export class UserService {
...
+    const editor = new UserEntity(userDummyEditor.username, hashString(userDummyEditor.password), Roles.EDITOR, true);
+    if (await this.isUsernameAlreadyInDatabase(editor.username)) {
+      console.log('Editor already in database.');
+    } else {
+      await this.userRepository.persistAndFlush(editor);
+    }
+  }
+
+  static async hashString(textToHash: string, rounds = 16) {
+    return await bcrypt.hash(textToHash, rounds);
```
Comment (lutangar) [inline apps/backend/src/user/user.service.ts]: You should add a pinch of salt here :
https://en.wikipedia.org/wiki/Salt_(cryptography)#:~:text=In%20cryptography%2C%20a%20salt%20is,to%20safeguard%20passwords%20in%20storage.

This salt could be kept as an env variable for example :
```
PASSWORD_SALT=areYouSalty?
```

And then :
```suggestion
    return await bcrypt.hash(textToHash+process.env.PASSWORD_SALT, rounds);
```

File: apps/backend/src/user/user.service.ts [hunk 5]
```diff
@@ -66,4 +68,26 @@ export class UserService {
...
+  ) {
+    const admin = new UserEntity(userAdmin.username, hashString(userAdmin.password), Roles.ADMIN, true);
+    if (await this.isUsernameAlreadyInDatabase(admin.username)) {
+      console.log('Admin already in database.');
+    } else {
+      await this.userRepository.persistAndFlush(admin);
+    }
+    const editor = new UserEntity(userDummyEditor.username, hashString(userDummyEditor.password), Roles.EDITOR, true);
+    if (await this.isUsernameAlreadyInDatabase(editor.username)) {
+      console.log('Editor already in database.');
```
Comment (lutangar) [inline apps/backend/src/user/user.service.ts]: Use `Logger` instead.

File: apps/backend/src/user/user.service.ts (hunk 1 above)
Comment (lutangar) [reply apps/backend/src/user/user.service.ts]: > see how to call static methods https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/static

File: apps/backend/src/project/project.controller.ts [hunk 6]
```diff
@@ -17,8 +18,12 @@ export class ProjectController {
   }
 
   @Post('project')
-  async create(@Body() project: ProjectEntity) {
-    return this.projectService.create(project);
+  async create(@Body() project: Pick<Project, 'title' | 'draft' | 'datasets' | 'description'>) {
+    // I stop here because I realize that I need the current user ID to create properly the project.
```
Comment (lutangar) [inline apps/backend/src/project/project.controller.ts:22]: It should be on another feature branch then.

**Review by lutangar — CHANGES_REQUESTED** [review]

File: apps/backend/src/user/user.service.ts (hunk 1 above)
Comment (racine-p-a) [reply apps/backend/src/user/user.service.ts]: As discussed : to be done at next rewriting of User stuff

**Review by Pipazoul — APPROVED** [review]

**Review by lutangar — APPROVED** [review]

**Pull request merged**